### PR TITLE
 ➕ Extra import fix of example

### PR DESCRIPTION
**Pull request by tkoyama010**
### Overview

Extra import fix of example. I found this in the sphinx compiling in local. We could not find the error by `pytest --doctest-modules` . Is there any good way to find this error in CI?

 

### Details

- Same with #1424

### CODE REVIEW

**Comment by akaszynski** [comment]
> Is there any good way to find this error in CI?

Issue here is the namespace is shared between docstrings.  I don't think there's a way getting around this, but I think @adeak might have looked into it.

Approving as CI appears to have run correctly.

**Review by akaszynski — APPROVED** [review]

**Comment by akaszynski** [comment]
Merging too since admin can override checks passing.

**Comment by tkoyama010** [comment]
@akaszynski Thanks

**Comment by adeak** [comment]
It seems the issue is deeper than I'd thought: it's that module globals are visible from doctests!

I've opened a discussion at https://github.com/pytest-dev/pytest/discussions/8791 after not being able to find a solution with an hour's digging.

**Comment by tkoyama010** [comment]
@adeak Thanks. Fantastic job!

**Comment by adeak** [comment]
So according to the pytest maintainer avoiding this is not supported, and not even really possible. Bummer.

I'm wondering how hard would it be to use stdlib `doctest` as a parser to do this on our own:
```py
import doctest
import pyvista as pv

# find and parse all docstrings in core.filters.poly_data
doctests = doctest.DocTestFinder().find(pv.core.filters.poly_data)
# 37 docstrings, out of which 25 have doctests (non-empty .example attr)

# find the first non-empty doctest
doct = next(dt for dt in doctests if dt.examples)
# <DocTest pyvista.core.filters.poly_data.PolyDataFilters.clean from /home/adeak/pyvista/pyvista/core/filters/poly_data.py:1131 (7 examples)>

# print each doctest row with expected output
for example in doct.examples:
    print(example.source.rstrip())
    if example.want:
        print(example.want)
```

This prints the following:
```
import pyvista as pv
import numpy as np
points = np.array([[0, 0, 0], [0, 1, 0], [1, 0, 0]])
faces = np.array([3, 0, 1, 2, 3, 0, 3, 3])
mesh = pv.PolyData(points, faces)
mout = mesh.clean()
print(mout.faces)  # doctest:+SKIP
[3 0 1 2]
```

We'd "only" have to
1. walk a tree of pyvista's structure, passing every interesting submodule into `DocTestFinder`
2. parse each doctest example by example, using an initially empty `globals()` dict
3. execute each line and comparing outputs, watching out for `doctest: +SKIP`s.

In other words, we'd only have to reimplement a skeleton of pytest :neutral_face: Perhaps we can just stick to checking missing imports, which would mean we can just `exec()` each example one after the other, and this would then raise. Whether the actual results are correct would be taken care of the doctesting of `pytest` or `sphinx`...

**Comment by adeak** [comment]
OK, here's another proof of concept because apparently that's my thing now:
```py
In [23]: import doctest
    ...: from functools import partial
    ...: from io import StringIO
    ...: import pyvista as pv
    ...: pv.OFF_SCREEN = True
    ...: 
    ...: # find and parse all docstrings in a submodule
    ...: doctests = doctest.DocTestFinder().find(pv.core.pointset, globs={})
    ...: 
    ...: for dt in doctests:
    ...:     if not dt.examples:
    ...:         continue
    ...:     globs = {'print': partial(print, file=StringIO())}
    ...:     for example in dt.examples:
    ...:         try:
    ...:             exec(example.source, globs)
    ...:         except Exception as e:
    ...:             print(f'FAILED: {dt}! ({e})')
    ...:             break
    ...:     else:
    ...:         print(f'PASSED: {dt}...')
PASSED: <DocTest pyvista.core.pointset.ExplicitStructuredGrid from /home/adeak/pyvista/pyvista/core/pointset.py:1122 (12 examples)>...
PASSED: <DocTest pyvista.core.pointset.ExplicitStructuredGrid.cast_to_unstructured_grid from /home/adeak/pyvista/pyvista/core/pointset.py:1236 (9 examples)>...
PASSED: <DocTest pyvista.core.pointset.ExplicitStructuredGrid.cell_coords from /home/adeak/pyvista/pyvista/core/pointset.py:1521 (4 examples)>...
PASSED: <DocTest pyvista.core.pointset.ExplicitStructuredGrid.cell_id from /home/adeak/pyvista/pyvista/core/pointset.py:1471 (5 examples)>...
PASSED: <DocTest pyvista.core.pointset.ExplicitStructuredGrid.compute_connections from /home/adeak/pyvista/pyvista/core/pointset.py:1735 (4 examples)>...
PASSED: <DocTest pyvista.core.pointset.ExplicitStructuredGrid.compute_connectivity from /home/adeak/pyvista/pyvista/core/pointset.py:1688 (4 examples)>...
PASSED: <DocTest pyvista.core.pointset.ExplicitStructuredGrid.dimensions from /home/adeak/pyvista/pyvista/core/pointset.py:None (3 examples)>...
PASSED: <DocTest pyvista.core.pointset.ExplicitStructuredGrid.hide_cells from /home/adeak/pyvista/pyvista/core/pointset.py:1325 (4 examples)>...
PASSED: <DocTest pyvista.core.pointset.ExplicitStructuredGrid.neighbors from /home/adeak/pyvista/pyvista/core/pointset.py:1564 (11 examples)>...
PASSED: <DocTest pyvista.core.pointset.ExplicitStructuredGrid.save from /home/adeak/pyvista/pyvista/core/pointset.py:1291 (9 examples)>...
PASSED: <DocTest pyvista.core.pointset.ExplicitStructuredGrid.show_cells from /home/adeak/pyvista/pyvista/core/pointset.py:1365 (6 examples)>...
PASSED: <DocTest pyvista.core.pointset.ExplicitStructuredGrid.visible_bounds from /home/adeak/pyvista/pyvista/core/pointset.py:None (5 examples)>...
PASSED: <DocTest pyvista.core.pointset.PointSet.remove_cells from /home/adeak/pyvista/pyvista/core/pointset.py:62 (3 examples)>...
PASSED: <DocTest pyvista.core.pointset.PolyData from /home/adeak/pyvista/pyvista/core/pointset.py:100 (14 examples)>...
PASSED: <DocTest pyvista.core.pointset.StructuredGrid from /home/adeak/pyvista/pyvista/core/pointset.py:908 (11 examples)>...
PASSED: <DocTest pyvista.core.pointset.StructuredGrid.hide_cells from /home/adeak/pyvista/pyvista/core/pointset.py:1071 (8 examples)>...
PASSED: <DocTest pyvista.core.pointset.UnstructuredGrid from /home/adeak/pyvista/pyvista/core/pointset.py:514 (7 examples)>...
FAILED: <DocTest pyvista.core.pointset.UnstructuredGrid._from_arrays from /home/adeak/pyvista/pyvista/core/pointset.py:638 (10 examples)>! (name 'np' is not defined)
PASSED: <DocTest pyvista.core.pointset.UnstructuredGrid.cast_to_explicit_structured_grid from /home/adeak/pyvista/pyvista/core/pointset.py:856 (9 examples)>...
```

(Note that penultimate failure.) And sure enough: https://github.com/pyvista/pyvista/blob/ec5aa343d857d0c7e6a79aeeba340797bc868ced/pyvista/core/pointset.py#L658-L665

(note the `np` vs `numpy` discrepancy).

I think we might even be able to get this to work. Several things to take care of still:
1. discovering submodules and classes/functions automatically
2. skipping doctest SKIP lines (easy)
3. _not_ skipping doctest SKIP lines at least for some local runs outside CI to weed out glaring errors hiding from CI?
4. reporting the results in some same way.

**Comment by akaszynski** [comment]
One consideration would be to simply add this script in our testing directory and call it separately in our CI.  Simply adding it in a PR is enough to have this part of our codebase and depending on the `pytest` developers, we may have to implement it as part of our regular checks.

I think running this from time to time outside of our normal CI might be fine too as a fun project if you want to keep doing that @adeak.

**Comment by adeak** [comment]
@akaszynski I've opened https://github.com/pyvista/pyvista/pull/1439 to start implementing this. Although it's a draft, I'd appreciate if you could take a look when your time permits, just to adjust the design if necessary. There's no rush at all, this is just a heads-up. I'm going to fix the few broken doctests it's found anyway.
